File: Hard/Median_of_Two_Sorted_Arrays/Median_of_Two_Sorted_Arrays.js
// Runtime: 104 ms, faster than 92.06% of JavaScript online submissions for Median of Two Sorted Arrays.
// Memory Usage: 39.2 MB, less than 72.34% of JavaScript online submissions for Median of Two Sorted Arrays.
// 02/14/2020
// MM/DD/YY

/*
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number}
*/

function merge(array1, array2) {
    let array = []
    let first = 0
    let second = 0

    while (first < array1.length && second < array2.length) {
        if (array1[first] > array2[second]) {
            array.push(array2[second])
			second++
		} else if (array1[first] < array2[second]) {
            array.push(array1[first])
			first++
		} else {
            array.push(array2[second])
            array.push(array1[first])
			second++
			first++
        }
    }

    while (first < array1.length) {
        array.push(array1[first])
        first++
    }

    while (second < array2.length) {
        array.push(array2[second])
        second++
    }

    return array
}

var findMedianSortedArrays = function(nums1, nums2) {
    let array = merge(nums1, nums2)
    if (array.length % 2 === 1) {
        return array[Math.floor(array.length/2)]
    } else {
        return (array[Math.floor(array.length/2)] + array[Math.floor(array.length/2)-1]) / 2
    }
}

File: Hard/Median_of_Two_Sorted_Arrays/Median_of_Two_Sorted_Arrays.py
#Runtime: 92 ms, faster than 84.09% of Python3 online submissions for Median of Two Sorted Arrays.
#Memory Usage: 12.7 MB, less than 100.00% of Python3 online submissions for Median of Two Sorted Arrays.
#02/14/2020
#MM/DD/YY

class Solution:
    #from merge sort
    @staticmethod
    def merge(array1, array2): #merge two arrays
        array = [] #merged array
        first = second = 0 #starting points

        while first < len(array1) and second < len(array2):
            if array1[first] > array2[second]:
                array.append(array2[second])
                second += 1
            elif array1[first] < array2[second]:
                array.append(array1[first])
                first += 1
            else:
                array.append(array1[first])
                array.append(array2[second])
                first += 1
                second += 1
        
        while first < len(array1):
            array.append(array1[first])
            first += 1

        while second < len(array2):
            array.append(array2[second])
            second += 1

        return array
    
    def findMedianSortedArrays(self, nums1: list, nums2: list) -> float:
        array = self.merge(nums1, nums2)
        if len(array) % 2 == 1:
            return float(array[len(array)//2])
        else:
            return (array[len(array)//2] + array[len(array)//2-1]) / 2